src/data/nat/fib/zeckendorf.lean

```diff
@@ -0,0 +1,229 @@
+  /- Hi : i = n + 2 -/
+  { rewrite Hi at *,
+    have no_n_plus_3 : n + 3 ∉ S,
+    { assume n_plus_3 : n + 3 ∈ S,
+      have bad : n + 3 ≤ n := Hle_n (n + 3) n_plus_3,
+      linarith },
+    finish },
+  /- Hi : i ∈ S -/
+  { obtain ⟨ge2, nconsec⟩ : i ≥ 2 ∧ i + 1 ∉ S := Hzeckendorf i Hi,
+    have not_new : i + 1 ≠ n + 2,
+    { have le_n : i ≤ n := Hle_n i Hi, linarith },
+    finish }
+end
+
+lemma zeckendorf_bound_base_case
+  (S : finset ℕ) (Hzeckendorf : is_zeckendorf_rep S)
+  (n : ℕ) (Hle_n : ∀ i ∈ S, i ≤ n)
+  (Hlt2 : n < 2):
+  S.sum fib < fib n.succ :=
+begin
+  suffices Hempty : S = ∅, by simp [Hempty, fib_pos],
+  apply finset.eq_empty_of_forall_not_mem,
+  assume (i : ℕ) (Hi : i ∈ S),
+  have ge_2: i ≥ 2 := (Hzeckendorf i Hi).elim_left,
+  have Hbad: i ≤ n := Hle_n i Hi,
+  linarith
+end
+
+/-- If we use Fibonacci numbers with indexes less than `n`, the sum will be less than `fib n`. -/
+lemma zeckendorf_bound
+  (S : finset ℕ) (Hzeckendorf : is_zeckendorf_rep S)
+  (n : ℕ) (Hle_n : ∀ i ∈ S, i ≤ n):
+  S.sum fib < fib n.succ :=
+begin
+  induction n using nat.two_step_induction with n IHn IHnsucc generalizing S,
+  { apply zeckendorf_bound_base_case; finish },
+  { apply zeckendorf_bound_base_case; finish },
+  /- Inductive case. Now we're bounded by n.succ.succ -/
+  { by_cases (n.succ.succ ∈ S),
+    /- Case 1: the bound is tight, i.e. `n.succ.succ ∈ S`. -/
+    { let S' := S.erase n.succ.succ,
+      have Hless_n_S' : ∀ i ∈ S', i ≤ n,
+      { assume (i : ℕ) (Hi : i ∈ S'),
+        obtain ⟨Hi_not_nsucc, Hi_in_S⟩: i ≠ n.succ.succ ∧ i ∈ S, by finish,
+        have less_n_succ : i ≤ n.succ.succ := Hle_n i Hi_in_S,
+        have Hi_not_n : i ≠ n.succ,
+        { assume (H: i = n.succ),
+          rw H at *,
+          have Hno_nsucc : n.succ.succ ∉ S := (Hzeckendorf n.succ Hi_in_S).elim_right,
+          contradiction },
+        show i ≤ n, by omega },
+      have Hzeckendorf' : is_zeckendorf_rep S',
+        { apply subset_is_zeckendorf Hzeckendorf, apply finset.erase_subset },
+      have S_insert : S = insert n.succ.succ S', by rwa finset.insert_erase,
+      have S'_no_nsucc : n.succ.succ ∉ S', by simp,
+      have IH := IHn S' Hzeckendorf' Hless_n_S',
+      calc ∑ k in S, fib k
+          = fib (n + 2) + S'.sum fib : by rwa [S_insert, finset.sum_insert]
+      ... < fib (n + 2) + fib (n + 1) : by simp [IHn S' Hzeckendorf' Hless_n_S']
+      ... = fib (n + 3) : by simp [@fib_add_two (n + 1)]; ring },
+    /- Case 2: the bound is not tight, i.e. `n.succ.succ ∉ S`. -/
+    { calc ∑ k in S, fib k < fib (n + 2) : _
+                       ... ≤ fib (n + 3) : by apply fib_mono; omega,
+      apply IHnsucc S Hzeckendorf,
+      assume (i : ℕ) (Hi : i ∈ S),
+      have less_n_succ : i ≤ n.succ.succ, from Hle_n i Hi,
+      have not_n : i ≠ n.succ.succ, by finish,
+      show i ≤ n.succ, by omega } }
+end
+
+theorem zeckendorf (n : ℕ): ∃! (S : finset ℕ), is_zeckendorf_rep S ∧ S.sum fib = n :=
```

Comment:
Would be better to have a more descriptive name and instead mention the name in the docstring.
```suggestion
/-- **Zeckendorf's theorem**. Natural numbers admit a unique representation in Fibonacci base. -/
theorem zeckendorf (n : ℕ): ∃! (S : finset ℕ), is_zeckendorf_rep S ∧ S.sum fib = n :=
```